vocone/vocone.go

```
@@ -0,0 +1,351 @@
+			oracleExist = true
+			break
+		}
+	}
+	if !oracleExist {
+		log.Warnf("adding new oracle key %s", oracleKey.Address())
+		vc.app.State.AddOracle(oracleKey.Address())
+		if _, err := vc.app.State.Save(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (vc *Vocone) setDefaultMethods() {
+	vc.app.IsSynchronizing = func() bool { return false }
+	vc.app.SetFnSendTx(vc.addTx)
+	vc.app.SetFnGetTx(vc.getTx)
+	vc.app.SetFnGetBlockByHeight(vc.getBlock)
+	vc.app.SetFnGetTxHash(vc.getTxWithHash)
+	vc.app.SetFnMempoolSize(vc.mempoolSize)
+}
+
+func (vc *Vocone) addTx(tx []byte) (*tmcoretypes.ResultBroadcastTx, error) {
+	var err error
+	resp := vc.app.CheckTx(abcitypes.RequestCheckTx{Tx: tx})
+	if resp.Code == 0 {
+		if err = vc.mempool.Enqueue(tx); err != nil {
+			return &tmcoretypes.ResultBroadcastTx{
+				Code: 1,
+				Data: []byte("mempool is full"),
+			}, fmt.Errorf("mempool is full")
+		}
+	} else {
+		log.Debugf("checkTx failed: %s", resp.Data)
+	}
+	return &tmcoretypes.ResultBroadcastTx{
+		Code: resp.Code,
+		Data: resp.Data,
+		Hash: ethereum.HashRaw(tx),
+	}, err
+}
+
+func (vc *Vocone) commitBlock() {
+	blockStoreTx := vc.blockStore.WriteTx()
+	var txCount int
+	for txCount = 0; txCount < vc.txsPerBlock; {
+		tx, err := vc.mempool.Dequeue()
+		if err != nil {
+			break
+		}
+		resp := vc.app.DeliverTx(abcitypes.RequestDeliverTx{Tx: tx.([]byte)})
+		if resp.Code == 0 {
+			blockStoreTx.Set(
+				[]byte(fmt.Sprintf("%d_%d", vc.height, txCount)),
+				tx.([]byte),
+			)
+			txCount++
+		} else {
+			log.Warnf("deliver tx failed: %s", resp.Data)
+		}
+	}
+	if txCount > 0 {
+		log.Infof("stored %d transactions on block %d", txCount, vc.height)
+		blockStoreTx.Commit()
+	} else {
+		blockStoreTx.Discard()
+	}
+}
+
+// TO-DO: improve this function
+func (vc *Vocone) getBlock(height int64) *tmtypes.Block {
+	blk := new(tmtypes.Block)
+	for i := int32(0); ; i++ {
+		tx, err := vc.getTx(uint32(height), i)
+		if err != nil {
+			break
```

ditto ignoring the error